test/model/typing_status_test.dart

```diff
@@ -0,0 +1,207 @@
+import 'package:flutter_test/flutter_test.dart';
+import 'package:zulip/api/model/events.dart';
+import 'package:zulip/api/model/model.dart';
+import 'package:zulip/model/narrow.dart';
+import 'package:zulip/model/typing_status.dart';
+
+import '../example_data.dart' as eg;
+import '../fake_async.dart';
+
+void main() {
+  late TypingStatus model;
+  late int notifiedCount;
+
+  void prepareModel() {
+    model = TypingStatus(
+      selfUserId: eg.selfUser.userId,
+      typingStartedExpiryPeriod: const Duration(milliseconds: 15000));
+    check(model.debugActiveNarrows).isEmpty();
+    notifiedCount = 0;
+    model.addListener(() => notifiedCount += 1);
+  }
+
+  void checkNotNotified() {
+    check(notifiedCount).equals(0);
+  }
+
+  void checkNotifiedOnce() {
+    check(notifiedCount).equals(1);
+    notifiedCount = 0;
+  }
+
+  void checkTypists(Map<SendableNarrow, List<User>> typistsByNarrow) {
+    final actualTypistsByNarrow = <SendableNarrow, Iterable<int>>{};
+    for (final narrow in model.debugActiveNarrows) {
+      actualTypistsByNarrow[narrow] = model.typistIdsInNarrow(narrow);
+    }
+    check(actualTypistsByNarrow).deepEquals(
+      typistsByNarrow.map((k, v) => MapEntry(k, v.map((e) => e.userId))));
+  }
+
+  final stream = eg.stream();
+  final topicNarrow = TopicNarrow(stream.streamId, 'foo');
+
+  final dmNarrow = DmNarrow.withUser(eg.otherUser.userId, selfUserId: eg.selfUser.userId);
+  final groupNarrow = DmNarrow.withOtherUsers(
+    [eg.otherUser.userId, eg.thirdUser.userId], selfUserId: eg.selfUser.userId);
+
+  group('handle typing start events', () {
+    test('add typists in separate narrows', () {
+      prepareModel();
+
+      model.handleTypingEvent(eg.typingEvent(dmNarrow, TypingOp.start, eg.otherUser.userId));
+      checkTypists({dmNarrow: [eg.otherUser]});
+      checkNotifiedOnce();
+
+      model.handleTypingEvent(eg.typingEvent(groupNarrow, TypingOp.start, eg.thirdUser.userId));
+      checkTypists({dmNarrow: [eg.otherUser], groupNarrow: [eg.thirdUser]});
+      checkNotifiedOnce();
+    });
+
+    test('add a typist in the same narrow', () {
+      prepareModel();
+
+      model.handleTypingEvent(eg.typingEvent(groupNarrow, TypingOp.start, eg.otherUser.userId));
+      checkTypists({groupNarrow: [eg.otherUser]});
+      checkNotifiedOnce();
+
+      model.handleTypingEvent(eg.typingEvent(groupNarrow, TypingOp.start, eg.thirdUser.userId));
+      checkTypists({groupNarrow: [eg.otherUser, eg.thirdUser]});
+      checkNotifiedOnce();
+    });
+
+    test('sort dm recipients', () {
+      prepareModel();
+
+      final eventUnsorted = TypingEvent(id: 1, op: TypingOp.start, senderId: eg.otherUser.userId,
+        messageType: MessageType.direct,
+        recipientIds: [5, 4, 10, 8, 2, 1, eg.selfUser.userId],
+        streamId: null,
+        topic: null);
+      // DmNarrow's constructor expects the recipient IDs to be sorted.
+      check(() => model.handleTypingEvent(eventUnsorted)).returnsNormally();
```

Comment:
Let's add a check that it has a typist at the expected DmNarrow. That's always a bit more reassuring than just "didn't throw" — if the code under test didn't throw, then maybe that's only because it hit an error but decided that meant it should do nothing, and it might not have even reached the bit of code the test was intended to exercise.

(These `returnsNormally` checks appear in some existing test cases you saw in the events tests, as baselines for comparison with `throws` checks. They work OK there, because their job is only to provide a contrast with the `throws` checks.)

Reply:
Yeah. I also found having to initialize the `DmNarrow` is informative in terms of where the sorted requirement comes from.